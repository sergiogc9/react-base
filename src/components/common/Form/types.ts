import React from 'react';
import * as yup from 'yup';
import { DeepMap, FieldError, UseFormProps } from 'react-hook-form';
import { BoxProps } from '@sergiogc9/react-ui';

export type FormHelpers = {
	setErrors: (errors: Record<string, string>) => void;
};

export type Props<FormValues extends Record<string, unknown> = Record<string, unknown>> = {
	readonly children: React.ReactNode;
	readonly defaultValues: Partial<FormValues>;
	readonly onChange?: (values: FormValues) => void;
	readonly onSubmit?: (values: FormValues, formHelpers: FormHelpers) => void;
	readonly onValidChange?: (isValid: boolean, errors: DeepMap<FormValues, FieldError>) => void;
	readonly useFormProps?: Omit<UseFormProps<FormValues>, 'defaultValues'>;
	readonly validationSchema?: yup.ObjectSchema<FormValues | undefined>;
};

export type FormProps<FormValues extends Record<string, unknown> = Record<string, unknown>> = Props<FormValues> &
	Omit<BoxProps<React.HTMLAttributes<HTMLFormElement>>, 'onChange' | 'onSubmit'>;

export type StyledFormProps = BoxProps<React.HTMLAttributes<HTMLFormElement>>;

export type FormEffectProps<FormValues extends Record<string, unknown> = Record<string, unknown>> = Pick<
	FormProps<FormValues>,
	'onChange' | 'onValidChange'
>;
