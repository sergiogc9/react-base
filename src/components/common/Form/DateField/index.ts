import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { State } from '@src/store/types';
import { DispatchProps, StateProps } from './types';
import FormDateField from './FormDateField';

const mapStateToProps = ({}: State): StateProps => ({});

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({});

export default connect(mapStateToProps, mapDispatchToProps)(FormDateField);
