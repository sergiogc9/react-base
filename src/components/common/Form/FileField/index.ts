import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { operators as notificationsOperators } from '@src/store/app/notifications';
import { DispatchProps, StateProps } from './types';
import FormFileField from './FormFileField';

const mapStateToProps = (): StateProps => ({});

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
	onAddNotification: notification => dispatch(notificationsOperators.add({notification}))
});

export default connect(mapStateToProps, mapDispatchToProps)(FormFileField);
