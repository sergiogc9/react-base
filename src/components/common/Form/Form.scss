@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.react-base-form {
	display: flex;
	flex-wrap: wrap;

	.react-base-form-field {
		flex-basis: 100%;
		padding: 0 1.5rem;
		position: relative;
		&.react-base-form-field-width-25 {
			flex-basis: 25%;
		}
		&.react-base-form-field-width-33 {
			flex-basis: 33%;
		}
		&.react-base-form-field-width-50 {
			flex-basis: 50%;
		}
		&.react-base-form-field-width-66 {
			flex-basis: 66%;
		}
		&.react-base-form-field-width-75 {
			flex-basis: 75%;
		}
		&.react-base-form-field-width-100 {
			flex-basis: 100%;
		}

		@include respond(tab-port) {
			flex-basis: 100% !important;
			padding: 0;
		}
	}

	.react-base-form-field-break-line {
		flex-basis: 100%;
		flex-grow: 1;
	}
}
