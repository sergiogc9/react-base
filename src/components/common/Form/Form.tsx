import React from 'react';
import { FormProvider, SubmitHandler, useForm, UseFormProps, useFormState, useWatch } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { isEmpty } from 'lib/imports/lodash';
import { useUpdateEffect } from '@sergiogc9/react-hooks';

import StyledForm from './styled';
import { FormEffectProps, FormProps } from './types';

// This component is needed to avoid fully re-render the whole Form component when form state changes.
const FormEffect = <FormValues extends Record<string, unknown>>(props: FormEffectProps<FormValues>) => {
	const { onChange, onValidChange } = props;

	const { errors, isValid, touchedFields } = useFormState<FormValues>();

	const values = useWatch<FormValues>({});

	useUpdateEffect(() => {
		if (onChange) onChange(values as FormValues);
	}, [values]);

	useUpdateEffect(() => {
		if (!isEmpty(touchedFields) && onValidChange) {
			onValidChange(isValid, errors);
		}
	}, [errors, touchedFields, isValid]);

	return null;
};

const Form = <FormValues extends Record<string, unknown>>(props: FormProps<FormValues>) => {
	const { children, defaultValues, onChange, onValidChange, onSubmit, validationSchema, useFormProps, ...rest } = props;

	const finalUseFormProps = React.useMemo<UseFormProps<FormValues>>(
		() => ({
			mode: 'onBlur',
			resolver: yupResolver(validationSchema),
			...useFormProps,
			defaultValues: defaultValues as any
		}),
		[defaultValues, useFormProps, validationSchema]
	);

	const methods = useForm(finalUseFormProps);
	const {
		formState: { isSubmitSuccessful },
		getValues,
		handleSubmit,
		reset,
		setError
	} = methods;

	React.useEffect(() => {
		if (isSubmitSuccessful) {
			reset(getValues() as any);
		}
	}, [reset, isSubmitSuccessful, getValues]);

	const setErrors = React.useCallback(
		(errors: Record<string, string>) => {
			Object.keys(errors).forEach(inputName =>
				setError(inputName as any, { type: 'manual', message: errors[inputName] })
			);
		},
		[setError]
	);

	const onFormSubmitted = React.useCallback<SubmitHandler<FormValues>>(
		async data => {
			if (onSubmit) await onSubmit(data as FormValues, { setErrors });
		},
		[onSubmit, setErrors]
	);

	return (
		<FormProvider {...methods}>
			<StyledForm {...rest} onSubmit={handleSubmit(onFormSubmitted)}>
				{children}
			</StyledForm>
			<FormEffect onChange={onChange} onValidChange={onValidChange} />
		</FormProvider>
	);
};

export default React.memo(Form) as typeof Form;
