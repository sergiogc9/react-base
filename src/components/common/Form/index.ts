import { createNameSpacedComponent } from '@sergiogc9/react-utils';

import FormButtonCancel from './Button/Cancel';
import FormButtonSubmit from './Button/Submit';
import FormCheckBox from './CheckBox';
import FormAppGoogleMapsAutocomplete from './AppGoogleMapsAutocomplete';
import FormTextField from './TextField';
import FormSelect from './Select';
import FormTextArea from './TextArea';
import FormSwitchBox from './SwitchBox';
import Form from './Form';

export type { FormErrors, FormHelpers, FormProps } from './types';
export default createNameSpacedComponent(Form, {
	AppGoogleMapsAutocomplete: FormAppGoogleMapsAutocomplete,
	ButtonCancel: FormButtonCancel,
	ButtonSubmit: FormButtonSubmit,
	CheckBox: FormCheckBox,
	Select: FormSelect,
	SwitchBox: FormSwitchBox,
	TextArea: FormTextArea,
	TextField: FormTextField
});
