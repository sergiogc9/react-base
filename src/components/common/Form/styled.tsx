import React from 'react';
import styled from 'styled-components';
import { Box } from '@sergiogc9/react-ui';

import { StyledFormProps } from './types';

const StyledForm: React.FC<StyledFormProps> = styled(Box)<StyledFormProps>``;

StyledForm.defaultProps = {
	as: 'form',
	display: 'block',
	width: '100%'
};

export default React.memo(StyledForm);
