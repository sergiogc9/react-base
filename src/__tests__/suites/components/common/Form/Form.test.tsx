import React from "react";
import { shallow } from "enzyme";
import { act } from 'react-dom/test-utils';

import Form from "@src/components/common/Form/Form";
import { ComponentProps } from "@src/components/common/Form/Form";
import TestHelper from "@src/__tests__/utils/Helper/TestHelper";

let onChangeMock = jest.fn();
describe("Form common component", () => {
	let wrapper: any;

	const updateComponent = (props: Partial<ComponentProps> = {}) => {
		act(() => {
			wrapper = shallow(
				<Form
					elements={[]}
					{...props}
				/>);
		});
	};

	beforeEach(() => {
		onChangeMock = jest.fn();
		updateComponent();
	});

	it("Component default snapshot", () => {
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a text field", () => {
		updateComponent({ elements: [TestHelper.textFieldElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a number field", () => {
		updateComponent({ elements: [TestHelper.numberFieldElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a select field", () => {
		updateComponent({ elements: [TestHelper.selectFieldElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a autocomplete field", () => {
		updateComponent({ elements: [TestHelper.autocompleteFieldElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a date field", () => {
		updateComponent({ elements: [TestHelper.dateFieldElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a file field", () => {
		updateComponent({ elements: [TestHelper.fileFieldelement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a component field", () => {
		updateComponent({ elements: [TestHelper.componentFormField] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a button group field", () => {
		updateComponent({ elements: [TestHelper.buttonGroupElement] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a filed with width style", () => {
		updateComponent({ elements: [{ ...TestHelper.fileFieldelement, style: { width: 50 } }] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot with a filed with unique in line style", () => {
		updateComponent({ elements: [{ ...TestHelper.fileFieldelement, style: { uniqueInLine: true } }] });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Component snapshot after set values", () => {
		updateComponent({ elements: [TestHelper.textFieldElement, TestHelper.numberFieldElement] });
		act(() => {
			wrapper.instance().setValue("textFieldId", "forcedValue");
			wrapper.instance().setValue("numberFieldId", 100);
		});
		expect(wrapper.debug()).toMatchSnapshot();
		expect(wrapper.instance().state.forceValues).toEqual({ textFieldId: "forcedValue", numberFieldId: 100 });
	});

	it("Component get values return correct values", () => {
		updateComponent({ elements: [TestHelper.textFieldElement, TestHelper.numberFieldElement] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "newVal", true);
		wrapper.find("Memo(FormNumberField)").at(0).props().onChangeNumber(TestHelper.numberFieldElement.id, 50, true);
		expect(wrapper.instance().getValues()).toEqual({ textFieldId: "newVal", numberFieldId: 50 });
	});


	it("Component get values validate form true", () => {
		updateComponent({ elements: [TestHelper.textFieldElement, TestHelper.numberFieldElement] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "newVal", true);
		wrapper.find("Memo(FormNumberField)").at(0).props().onChangeNumber(TestHelper.numberFieldElement.id, 50, true);
		expect(wrapper.instance().validate()).toEqual(true);
	});

	it("Component validate form false", () => {
		updateComponent({ elements: [TestHelper.textFieldElement, TestHelper.numberFieldElement] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "newVal", true);
		wrapper.find("Memo(FormNumberField)").at(0).props().onChangeNumber(TestHelper.numberFieldElement.id, 50, false);
		expect(wrapper.instance().validate()).toEqual(false);
	});

	it("Component validate form with required elements returns false", () => {
		updateComponent({ elements: [TestHelper.textFieldElement, { ...TestHelper.numberFieldElement, required: true }] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "newVal", true);
		expect(wrapper.instance().validate()).toEqual(false);
	});

	it("Component validate form with required text input returns false", () => {
		updateComponent({ elements: [{ ...TestHelper.textFieldElement, required: true }, TestHelper.numberFieldElement] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "", true);
		expect(wrapper.instance().validate()).toEqual(false);
	});

	it("Component on element changed should call onValueUpdated if exists", () => {
		const onValueMock = jest.fn();
		updateComponent({ elements: [{ ...TestHelper.textFieldElement, onValueUpdated: onValueMock }] });
		wrapper.find("Memo(FormTextField)").at(0).props().onChangeText(TestHelper.textFieldElement.id, "newVal", true);
		expect(onValueMock).toHaveBeenCalledTimes(1);
	});
});
