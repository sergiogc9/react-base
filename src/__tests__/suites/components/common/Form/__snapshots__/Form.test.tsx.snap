// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form common component Component default snapshot 1`] = `"<div className=\\"react-base-form\\" />"`;

exports[`Form common component Component snapshot after set values 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormTextField) element={{...}} forceValue=\\"forcedValue\\" error={false} onChangeText={[Function]} t={[Function: t]} />
  </div>
  <div className=\\"react-base-form-field\\">
    <Memo(FormNumberField) element={{...}} forceValue={100} error={false} onChangeNumber={[Function]} t={[Function: t]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a autocomplete field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormAutocompleteField) element={{...}} forceValue={[undefined]} error={false} onAutocompleteSelected={[Function]} onUpdateItems={[undefined]} t={[Function: t]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a button group field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormButtonGroupField) element={{...}} forceValue={[undefined]} onChangeOption={[Function]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a component field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <div />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a date field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Connect(Component) element={{...}} forceValue={[undefined]} error={false} onChangeDate={[Function]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a file field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Connect(Component) element={{...}} forceValue={[undefined]} error={false} onChangeFile={[Function]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a filed with unique in line style 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field-break-line\\" />
  <div className=\\"react-base-form-field\\">
    <Connect(Component) element={{...}} forceValue={[undefined]} error={false} onChangeFile={[Function]} />
  </div>
  <div className=\\"react-base-form-field-break-line\\" />
</div>"
`;

exports[`Form common component Component snapshot with a filed with width style 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field react-base-form-field-width-50\\">
    <Connect(Component) element={{...}} forceValue={[undefined]} error={false} onChangeFile={[Function]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a number field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormNumberField) element={{...}} forceValue={[undefined]} error={false} onChangeNumber={[Function]} t={[Function: t]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a select field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormSelectField) element={{...}} forceValue={[undefined]} error={false} onSelectOption={[Function]} t={[Function: t]} />
  </div>
</div>"
`;

exports[`Form common component Component snapshot with a text field 1`] = `
"<div className=\\"react-base-form\\">
  <div className=\\"react-base-form-field\\">
    <Memo(FormTextField) element={{...}} forceValue={[undefined]} error={false} onChangeText={[Function]} t={[Function: t]} />
  </div>
</div>"
`;
